File: ios/EntrigModule.swift
import React
import UserNotifications
import EntrigSDK

@objc(Entrig)
class EntrigModule: RCTEventEmitter {
  private var hasListeners = false

  override func supportedEvents() -> [String]! {
    return ["onForegroundNotification", "onNotificationOpened"]
  }

  override class func requiresMainQueueSetup() -> Bool {
    return true
  }

  override func startObserving() {
    hasListeners = true
  }

  override func stopObserving() {
    hasListeners = false
  }

  @objc(initialize:withResolver:withRejecter:)
  func initializeSDK(config: [String: Any], resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let apiKey = config["apiKey"] as? String, !apiKey.isEmpty else {
      rejecter("INVALID_API_KEY", "API key is required and cannot be empty", nil)
      return
    }

    let handlePermission = config["handlePermission"] as? Bool ?? true
    let showForegroundNotification = config["showForegroundNotification"] as? Bool ?? true
    let entrigConfig = EntrigConfig(apiKey: apiKey, handlePermission: handlePermission, showForegroundNotification: showForegroundNotification)

    // Set up SDK listeners
    Entrig.setOnForegroundNotificationListener(self)
    Entrig.setOnNotificationOpenedListener(self)

    Entrig.configure(config: entrigConfig) { success, error in
      if success {
        resolver(nil)
      } else {
        rejecter("INIT_ERROR", error ?? "Failed to initialize SDK", nil)
      }
    }
  }

  @objc(register:withIsDebug:withResolver:withRejecter:)
  func register(userId: String, isDebugOverride: NSNumber?, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    let isDebug: Bool
    if let override = isDebugOverride {
      isDebug = override.boolValue
    } else {
      #if DEBUG
      isDebug = true
      #else
      isDebug = false
      #endif
    }

    Entrig.register(userId: userId, sdk: "react-native", isDebug: isDebug) { success, error in
      if success {
        resolver(nil)
      } else {
        rejecter("REGISTER_ERROR", error ?? "Registration failed", nil)
      }
    }
  }

  @objc(requestPermission:withRejecter:)
  func requestPermission(resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    Entrig.requestPermission { granted, error in
      if let error = error {
        rejecter("PERMISSION_ERROR", error.localizedDescription, nil)
      } else {
        resolver(granted)
      }
    }
  }

  @objc(unregister:withRejecter:)
  func unregister(resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    Entrig.unregister { success, error in
      if success {
        resolver(nil)
      } else {
        rejecter("UNREGISTER_ERROR", error ?? "Unregistration failed", nil)
      }
    }
  }

  @objc(getInitialNotification:withRejecter:)
  func getInitialNotification(resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    if let event = Entrig.getInitialNotification() {
      let payload: [String: Any] = [
        "title": event.title ?? "",
        "body": event.body ?? "",
        "data": event.data ?? [:],
        "isForeground": false
      ]
      resolver(payload)
    } else {
      resolver(nil)
    }
  }

  // Helper method to send notification events to JavaScript
  private func sendNotificationToJS(event: NotificationEvent, isForeground: Bool) {
    guard hasListeners else { return }

    let payload: [String: Any] = [
      "title": event.title ?? "",
      "body": event.body ?? "",
      "data": event.data ?? [:],
      "isForeground": isForeground
    ]

    if isForeground {
      sendEvent(withName: "onForegroundNotification", body: payload)
    } else {
      sendEvent(withName: "onNotificationOpened", body: payload)
    }
  }
}

// MARK: - SDK Listeners
extension EntrigModule: OnNotificationReceivedListener {
  func onNotificationReceived(_ event: NotificationEvent) {
    sendNotificationToJS(event: event, isForeground: true)
  }
}

extension EntrigModule: OnNotificationClickListener {
  func onNotificationClick(_ event: NotificationEvent) {
    sendNotificationToJS(event: event, isForeground: false)
  }
}

File: android/src/main/java/com/entrig/reactnative/EntrigModule.kt
package com.entrig.reactnative

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.activity.ComponentActivity
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.entrig.sdk.Entrig
import com.entrig.sdk.models.EntrigConfig

class EntrigModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext), ActivityEventListener {

  private val context: Context
    get() = reactApplicationContext

  override fun getName(): String = "Entrig"

  override fun initialize() {
    super.initialize()

    // Register for activity events (onNewIntent, onActivityResult)
    reactApplicationContext.addActivityEventListener(this)

    val currentActivity = reactApplicationContext.currentActivity
    // Set activity on SDK for foreground detection
    currentActivity?.let { Entrig.setActivity(it) }

    // Handle initial intent (app launched from notification tap)
    currentActivity?.intent?.let { intent ->
      Entrig.handleIntent(intent)
    }

    Entrig.setOnForegroundNotificationListener { notification ->
      sendEvent("onForegroundNotification", Arguments.makeNativeMap(notification.toMap() as Map<String, Any?>))
    }

    Entrig.setOnNotificationOpenedListener { notification ->
      sendEvent("onNotificationOpened", Arguments.makeNativeMap(notification.toMap() as Map<String, Any?>))
    }
  }

  override fun invalidate() {
    reactApplicationContext.removeActivityEventListener(this)
    super.invalidate()
  }

  // ActivityEventListener — called when a new intent arrives (notification tap while app is open)
  override fun onNewIntent(intent: Intent) {
    Entrig.handleIntent(intent)
  }

  override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
    // Not used
  }

  @ReactMethod
  fun initialize(config: ReadableMap, promise: Promise) {
    val apiKey = config.getString("apiKey")
    if (apiKey.isNullOrEmpty()) {
      promise.reject("INVALID_API_KEY", "API key is required and cannot be empty")
      return
    }

    val showForegroundNotification = if (config.hasKey("showForegroundNotification")) {
      config.getBoolean("showForegroundNotification")
    } else {
      true
    }
    val entrigConfig = EntrigConfig(
      apiKey = apiKey,
      handlePermission = false, // Module handles permission itself
      showForegroundNotification = showForegroundNotification
    )

    val appCtx = context.applicationContext
    Entrig.initialize(appCtx, entrigConfig) { success, error ->
      if (success) {
        promise.resolve(null)
      } else {
        promise.reject("INIT_ERROR", error ?: "Failed to initialize SDK")
      }
    }
  }

  @ReactMethod
  fun register(userId: String, isDebug: Boolean?, promise: Promise) {
    val activity = reactApplicationContext.currentActivity
    if (activity == null) {
      promise.reject("NO_ACTIVITY", "Activity not available")
      return
    }

    if (needsNotificationPermission()) {
      requestNotificationPermission(activity) {
        // Proceed with registration regardless of permission result
        doRegister(userId, activity, promise)
      }
      return
    }

    doRegister(userId, activity, promise)
  }

  @ReactMethod
  fun requestPermission(promise: Promise) {
    val activity = reactApplicationContext.currentActivity
    if (activity == null) {
      promise.reject("NO_ACTIVITY", "Activity not available")
      return
    }

    if (needsNotificationPermission()) {
      requestNotificationPermission(activity) { granted ->
        promise.resolve(granted)
      }
      return
    }

    promise.resolve(true)
  }

  @ReactMethod
  fun unregister(promise: Promise) {
    Entrig.unregister { success, error ->
      if (success) {
        promise.resolve(null)
      } else {
        promise.reject("UNREGISTER_ERROR", error ?: "Unregistration failed")
      }
    }
  }

  @ReactMethod
  fun getInitialNotification(promise: Promise) {
    val initialNotification = Entrig.getInitialNotification()
    if (initialNotification != null) {
      promise.resolve(Arguments.makeNativeMap(initialNotification.toMap() as Map<String, Any?>))
    } else {
      promise.resolve(null)
    }
  }

  private fun needsNotificationPermission(): Boolean {
    return Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
      ContextCompat.checkSelfPermission(context, android.Manifest.permission.POST_NOTIFICATIONS) !=
        PackageManager.PERMISSION_GRANTED
  }

  private fun requestNotificationPermission(activity: Activity, callback: (Boolean) -> Unit) {
    val componentActivity = activity as? ComponentActivity
    if (componentActivity == null) {
      callback(false)
      return
    }

    val key = "entrig_permission_${System.nanoTime()}"
    val registry = componentActivity.activityResultRegistry

    var launcher: androidx.activity.result.ActivityResultLauncher<String>? = null
    launcher = registry.register(key, ActivityResultContracts.RequestPermission()) { granted ->
      launcher?.unregister()
      callback(granted)
    }
    launcher.launch(android.Manifest.permission.POST_NOTIFICATIONS)
  }

  private fun doRegister(userId: String, activity: Activity, promise: Promise) {
    Entrig.register(userId, activity, "react-native") { success, error ->
      if (success) {
        promise.resolve(null)
      } else {
        promise.reject("REGISTER_ERROR", error ?: "Registration failed")
      }
    }
  }

  private fun sendEvent(eventName: String, params: WritableMap?) {
    if (reactApplicationContext.hasActiveReactInstance()) {
      reactApplicationContext
        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
        .emit(eventName, params)
    }
  }
}
